### ISSUE TITLE
Tracking Issue for XXX

### ISSUE THREAD

**Issue by EliasHolzmann** [rust-lang/rust#118117]
Feature gate: `#![feature(formatting_options)]`

This is a tracking issue for `fmt::FormattingOptions`.

`FormattingOptions` can be used to construct custom `Formatter`s at runtime.

### Public API

```rust
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct FormattingOptions { … }

enum Sign {
    Plus, 
    Minus
}

impl FormattingOptions {
    pub fn new() -> Self;

    pub fn sign(&mut self, sign: Option<Sign>) -> &mut Self;
    pub fn zero_pad(&mut self, zero_pad: bool) -> &mut Self;
    pub fn alternate(&mut self, alternate: bool) -> &mut Self;
    pub fn fill(&mut self, fill: Option<char>) -> &mut Self;
    pub fn alignment(&mut self, alignment: Option<Alignment>) -> &mut Self;
    pub fn width(&mut self, width: Option<usize>) -> &mut Self;
    pub fn precision(&mut self, precision: Option<usize>) -> &mut Self;

    pub fn get_sign(&self) -> Option<Sign>;
    pub fn get_zero_pad(&self) -> bool;
    pub fn get_alternate(&self) -> bool;
    pub fn get_fill(&self) -> Option<char>;
    pub fn get_alignment(&self) -> Option<Alignment>;
    pub fn get_width(&self) -> Option<usize>;
    pub fn get_precision(&self) -> Option<usize>;
    
    pub fn create_formatter<'a>(self, write: &'a mut (dyn Write + 'a)) -> Formatter<'a>;
}

impl<'a> Formatter<'a> {
    pub fn new(write: &'a mut (dyn Write + 'a), options: FormattingOptions) -> Self;
    pub fn with_options(&mut self, options: FormattingOptions) -> Self;
    
    pub fn sign(&self) -> Option<Sign>;

    pub fn options(&self) -> FormattingOptions;
}
```

### Steps / History

- [x] API change proposal (ACP)[^1]: https://github.com/rust-lang/libs-team/issues/286
- [ ] Implementation: #TBD
- [ ] Final comment period (FCP)[^2]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html
[^2]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by joboet** [comment]
I was wondering: could the signature for `Formatter::new` be changed to take a reference to the `FormattingOptions` instead of passing them by value? That way, LLVM wouldn't need to copy around the options as often, which seems like it could improve performance and binary size in cases where the same options are used multiple times, for instance when using the default options.

**Comment by EliasHolzmann** [comment]
@joboet I don't think so. `Formatter` directly holds a `FormattingOptions` internally: https://doc.rust-lang.org/nightly/src/core/fmt/mod.rs.html#524

If `Formatter::new` takes its `FormattingOptions` by reference instead of by value, we'd have to copy the options in that function anyway. (Instead, you probably could have the contents of `FormattingOptions` behind a `Rc` so that copying/cloning is cheaper, which might indeed improve performance for use cases with heavy `FormattingOptions` reuse. However, I don't think that this will compensate the ref-counting overhead that is in most use cases probably unnecessary.)

In most cases, when the `FormattingOptions` can be used multiple times, the target buffer of the resulting `Formatter` is probably also the same – in this case, you can just use the `Formatter` you construct.

**Comment by Ternvein** [comment]
As far as I can see, `&mut T` in `const` contexts are now stable. Should `const` be added for functions with mutable references?

**Comment by EliasHolzmann** [comment]
@Ternvein Thanks for the notice, didn't see that! I'll create a PR for that.

---

@TechPizzaDev:
> On the topic of references, shouldn't the options getter return a reference instead of a value? Leave it up to the user to clone it.

In many cases, having getters return references is probably a good idea. However, in this instance, I disagree. All the getters return data types that are `Copy` and aren't bigger than pointer size, I think. Having the getters return references that are later dereferenced by the user would be bad for performance (unless the compiler optimizes the unnecessary passing of the pointer away).

> Feels like `fn into_options(self) -> FormattingOptions` could exist for consistency.

As a method on `Formatter`, as an alternative for `fn options(&self) -> FormattingOptions`? I don't think duplicating the functionality is desirable. Also, passing `self` by value would have overhead compared to passing a reference, as a `Formatter` is relatively big (`FormattingOptions` plus a fat pointer for the `buf` trait object reference).

**Comment by TechPizzaDev** [comment]
`FormattingOptions` was 3x pointer size last I checked, which is why I found the `Copy` derive strange.

There was discussion to pack it down, but that would be annoying if any new fields were to be added in the future.

**Comment by vitu168** [comment]
Hello
